src/compute-client/src/controller/instance.rs

```diff
@@ -1097,88 +1082,39 @@ where
         // Validate the dataflow as having inputs whose `since` is less or equal to the dataflow's `as_of`.
         // Start tracking frontiers for each dataflow, using its `as_of` for each index and sink.
 
-        // When we initialize per-replica input frontiers (and thereby the per-replica read
-        // capabilities), we cannot use the `as_of` because of reconciliation: Existing
-        // slow replicas might be reading from the inputs at times before the `as_of` and we
-        // would rather not crash them by allowing their inputs to compact too far. So instead
-        // we initialize the per-replica write frontiers with the smallest possible value that
-        // is a valid read capability for all inputs, which is the join of all input `since`s.
-        let mut replica_input_frontier = Antichain::from_elem(T::minimum());
-
-        // Record all transitive dependencies of the outputs.
-        let mut storage_dependencies = Vec::new();
-        let mut compute_dependencies = Vec::new();
-
-        // Any potentially acquired STORAGE read holds. We acquire them and
-        // check whether our as_of is valid. They are dropped once we installed
-        // read capabilities manually.
-        //
-        // TODO: Instead of acquiring these and then dropping later, we should
-        // instead store them and don't "manually" acquire read holds using
-        // `update_read_capabilities`.
-        let mut storage_read_holds = Vec::new();
+        // When we install per-replica input read holds, we cannot use the `as_of` because of
+        // reconciliation: Existing slow replicas might be reading from the inputs at times before
+        // the `as_of` and we would rather not crash them by allowing their inputs to compact too
+        // far. So instead we take read holds at the least time available.
```

Comment:
I didn't follow this comment. Don't we get feedback from clusters about their read frontiers? I would expect that we can get a precise read hold by looking at what was the last read frontier update we last received across all replicas. Also, how do we determine the "least time available"? Afaik persist doesn't have a guaranteed way of giving you the global since of a shard

Reply:
The problematic scenario is when the controller just starts up and hasn't yet reconciled with the replicas. At that point it hasn't received frontier information from the replicas yet, so we can't assume anything except that the previous instance of the controller did have read holds sufficiently far in the past for what's required by the existing replicas.

> Also, how do we determine the "least time available"?

This is just the least-downgraded read hold we can get from the storage controller. It's fundamentally best effort. If storage decides to give out a read hold that is downgraded too far, there is nothing we can do except hope that the existing replicas' reader leases don't expire.

> Afaik persist doesn't have a guaranteed way of giving you the global since of a shard

I think during bootstrap we are guaranteed to initialize the read frontiers of storage collections with the current persist sinces? If we are not then I don't see how any of the bootstrap as-of selection could work. For example, an MV dataflow needs to continue where it left off, but if we would initialize storage collections to arbitrary values greater than the persist sinces, that wouldn't always be possible.

Reply:
> I think during bootstrap we are guaranteed to initialize the read frontiers of storage collections with the current persist sinces? If we are not then I don't see how any of the bootstrap as-of selection could work.

We are initializing them to the SinceHandle of the controller but I was thinking that a slow replica could hold onto a leased reader that is further behind the controller's read handle, since the latter is updated through movement of the upper, which follows the faster replica.